src/NServiceBus.Core/Recoverability/Settings/DelayedRetriesSettings.cs

```diff
@@ -36,5 +39,42 @@ public DelayedRetriesSettings TimeIncrease(TimeSpan timeIncrease)
 
             return this;
         }
+
+        /// <summary>
+        /// Registers a callback which is invoked when a message fails processing and will be retried after a delay.
+        /// </summary>
+        public DelayedRetriesSettings OnMessageBeingRetried(Action<DelayedRetryMessage> notificationCallback)
```

Comment:
I personally prefer the notation without `On` but that is just my taste.

Reply:
I think the `On` helps understand that this is related to an event and makes it easier to grasp the nature of the API compared to `settings.MessageBeingRetried(x => ...)`.  Can you help me understand why you prefer to skip the `On`?

Reply:
Coming from the angle that most of the types that we exposed are directly related to the intent they are having like `IHandleMessages` where the `I` is not really the indicator of an interface but more of a fluent descriptor of what the class that implements it does. Another one that makes me lean towards the other one is that the method itself is not invoked by us like an event handler is but the actual action delegate `notificationCallback` that you pass in

Reply:
I'm not really seeing a direct correlation between this API and the `IDoXYZ` APIs (which are arguably differ quite a bit from general .NET naming conventions).

> Another one that makes me lean towards the other one is that the method itself is not invoked by us like an event handler is but the actual action delegate

afaik, it's quite common to name methods that accept the callback for an event exactly this way. "On this event, execute this callback".

I definitely prefer the `On` prefix over just `MessageBeingRetried`. There might be other naming options though? Happy to schedule a call if you want to discuss this further.

Reply:
👍 for `On` -- this is the language I am used to.

Reply:
The `On` convention was introduced for methods that fire event handlers, see https://docs.microsoft.com/en-us/previous-versions/visualstudio/visual-studio-2008/wkzf914z(v=vs.90)?redirectedfrom=MSDN

My point was, in general, our public APIs do not add any type of Prefix or Postfix. So why is on needed here? Even if you look at how the previous API was named it didn't contain any Prefix or Postfix. Using `MessageBeingRetried(Action<ImmediateRetryMessage> notificationCallback)` or `MessageBeingRetried(Action<ImmediateRetryMessage> on)` or `MessageBeingRetried(Action<ImmediateRetryMessage> callback)` is expressive enough and the `On` prefix does not add any value.

Reply:
the linked documentation also states:
> The protected OnEventNamemethod also allows derived classes to override the event without attaching a delegate to it.

which is imo more relvant to this case as we're not dealing with .NET events and delegates here.

I prefer the `On` prefix. I don't see this causing any issues and potentially easier to understand. I could see `MessageBeingRetriedCallback` as an alternative though.

Reply:
I did some research in .NET. There is one example that I found that uses the `On` naming convention

https://referencesource.microsoft.com/#mscorlib/system/runtime/compilerservices/INotifyCompletion.cs,28

Other methods that accept an action delegate are named after what it does. For example 

`CancellationToken.Register(Action callback)` indicates that you can register multiple callbacks. 

Other examples are `Parallel.Invoke(Action action)` where Invoke is used to convey that the method provided will be immediately invoked. `Task.Run(Action action)` is also formulated from the perspective of the tasks.

Reflecting on that I was about to propose `RegisterXYZ` or just `Register(Action<Payload> messageBeingRetriedCallback)` or something similar

Reply:
Anyway, I don't want to annoy you any further.

Reply:
This seems to focus more on the naming convention than the maintenance release.
May I suggest raising an issue (architecture?) to tackle this separately and to be applied throughout the code base, rather than just a single spot.

Reply:
We found https://github.com/Particular/NServiceBus/blob/ead779f33e8bdec6844ee99a892729cfd7b7f0bc/src/NServiceBus.Core/Pipeline/Incoming/ReceivePipelineConfigExtensions.cs#L15 and came to the conclusion that the `On` prefix is definitely a convention that we want to stick to.